Controllers/Auth.php

```diff
@@ -0,0 +1,920 @@
+				// redirect them back to the admin page
+				$this->session->setFlashdata('message', $this->ionAuth->messages());
+				return redirect('/auth');
+			}
+		}
+		else
+		{
+			// display the create group form
+			// set the flash data error message if there is one
+			$this->data['message'] = ($this->form_validation->listErrors() ? $this->form_validation->listErrors() : ($this->ionAuth->errors() ? $this->ionAuth->errors() : $this->session->getFlashdata('message')));
+
+			$this->data['group_name'] = [
+				'name'  => 'group_name',
+				'id'    => 'group_name',
+				'type'  => 'text',
+				'value' => set_value('group_name'),
+			];
+			$this->data['description'] = [
+				'name'  => 'description',
+				'id'    => 'description',
+				'type'  => 'text',
+				'value' => set_value('description'),
+			];
+
+			return $this->_render_page('auth/create_group', $this->data);
+		}
+	}
+
+	/**
+	 * Edit a group
+	 *
+	 * @param int|string $id
+	 */
+	public function edit_group($id)
+	{
+		// bail if no group id given
+		if (!$id || empty($id))
+		{
+			return redirect('/auth');
+		}
+
+		$this->data['title'] = lang('Auth.edit_group_title');
+
+		if (!$this->ionAuth->logged_in() || !$this->ionAuth->is_admin())
+		{
+			return redirect('/auth');
+		}
+
+		$group = $this->ionAuth->group($id)->row();
+
+		// validate form input
+		$this->form_validation->setRule('group_name', lang('Auth.edit_group_validation_name_label'), 'required|alpha_dash');
+
+		if (isset($_POST) && !empty($_POST))
+		{
+			if ($this->form_validation->withRequest($this->request)->run())
+			{
+				$group_update = $this->ionAuth->update_group($id, $_POST['group_name'], $_POST['group_description']);
+
+				if ($group_update)
+				{
+					$this->session->setFlashdata('message', lang('Auth.edit_group_saved'));
+				}
+				else
+				{
+					$this->session->setFlashdata('message', $this->ionAuth->errors());
+				}
+				return redirect('/auth');
+			}
+		}
+
+		// set the flash data error message if there is one
+		$this->data['message'] = ($this->form_validation->listErrors() ? $this->form_validation->listErrors() : ($this->ionAuth->errors() ? $this->ionAuth->errors() : $this->session->getFlashdata('message')));
+
+		// pass the user to the view
+		$this->data['group'] = $group;
+
+		$readonly = $this->configIonAuth->admin_group === $group->name ? 'readonly' : '';
+
+		$this->data['group_name'] = [
+			'name'    => 'group_name',
+			'id'      => 'group_name',
+			'type'    => 'text',
+			'value'   => set_value('group_name', $group->name),
+			$readonly => $readonly,
+		];
+		$this->data['group_description'] = [
+			'name'  => 'group_description',
+			'id'    => 'group_description',
+			'type'  => 'text',
+			'value' => set_value('group_description', $group->description),
+		];
+
+		return $this->_render_page('auth' . DIRECTORY_SEPARATOR . 'edit_group', $this->data);
+	}
+
+	/**
+	 * @return array A CSRF key-value pair
+	 */
+	public function _get_csrf_nonce(): array
```

Comment:
I think this can be refactored away now that CI offerers CSRF built in

Reply:
You 'are right.
We can remove CSRF in the branch '3' to.